core/src/apps/cardano/address.py

```diff
@@ -34,6 +39,39 @@
 MAX_ADDRESS_BYTES_LENGTH = 65
 
 
+def validate_address_parameters(parameters: CardanoAddressParametersType) -> None:
+    if parameters.address_type == CardanoAddressType.BYRON:
+        if not is_byron_path(parameters.address_n):
+            raise INVALID_ADDRESS_PARAMETERS
+    elif parameters.address_type in ADDRESS_TYPES_SHELLEY:
+        if not is_shelley_path(parameters.address_n):
+            raise INVALID_ADDRESS_PARAMETERS
+
+        if parameters.address_type == CardanoAddressType.BASE:
+            _validate_base_address_staking_info(
+                parameters.address_n_staking, parameters.staking_key_hash
+            )
+        elif parameters.address_type == CardanoAddressType.POINTER:
+            if parameters.certificate_pointer is None:
+                raise INVALID_ADDRESS_PARAMETERS
+        elif parameters.address_type == CardanoAddressType.REWARD:
+            if not SCHEMA_STAKING_ANY_ACCOUNT.match(parameters.address_n):
+                raise INVALID_ADDRESS_PARAMETERS
+    else:
+        raise INVALID_ADDRESS_PARAMETERS
+
+
+def _validate_base_address_staking_info(
+    staking_path: list[int],
+    staking_key_hash: bytes | None,
+) -> None:
+    if (staking_key_hash is None) == (not staking_path):
+        raise INVALID_ADDRESS_PARAMETERS
+
+    if staking_key_hash is None and not SCHEMA_STAKING_ANY_ACCOUNT.match(staking_path):
```

Comment:
I'd find it more natural to check `staking_path is not None and ...` - the mutual exclusive presence of those parameters is being checked by the top assertion after all

Reply:
See [comment](https://github.com/vacuumlabs/trezor-firmware/pull/60#discussion_r617306449) above.